### PR TITLE
CircleCI custom builds

### PR DESCRIPTION
**Pull request by makarandp0**
CircleCI does not have built-in UI to launch builds manually. But it does provide APIs for it. This change enables us to request a custom build from CircleCI. Custom build is an integration tests run where caller can configure 
- browser (chrome,firefox)
- bver (stable, beta, unstable)
- environment (prod, stage, dev)
- branch 

 
To launch a custom build, execute the command `.circleci/run_on_circle.js` 
it will prompt ask bunch of questions and queue a build.
```
$ .circleci/run_on_circle.js 
? Circle CI Token ? forbar
? Environment? stage
? Workflow? custom
? Branch? circle_custom_builds
? Browser? firefox
? Bver ? unstable
? Topology ? peer-to-peer

```

Example: https://app.circleci.com/jobs/github/twilio/twilio-video.js/1372

I am hoping that this would be useful for a quick builds, and also for other teams/jobs to request our circleCI runs. like our jenkins integration test job that runs on Jenkins. 
 
**Contributing to Twilio**

> All third-party contributors acknowledge that any contributions they provide will be made under the same open-source license that the open-source project is provided under.

- [x] I acknowledge that all my contributions will be made under the project's license.

### CODE REVIEW

File: .circleci/run_on_circle.js [hunk 1]
```diff
@@ -0,0 +1,167 @@
+#!/usr/bin/env node
```
Comment (manjeshbhargav) [inline .circleci/run_on_circle.js]: @makarandp0 Can the filename be something like: `runbuild.js`? We normally don't use `_`s in our filenames.

File: .circleci/run_on_circle.js [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    body.parameters.bver = program.bver;
+    body.parameters.topology = program.topology;
+  }
+
+  return { options, body };
+}
+
+// sends a request using given options/body
+function triggerBuild({options, body}) {
+  console.log('options:', options);
```
Comment (manjeshbhargav) [inline .circleci/run_on_circle.js]:
```suggestion
```

File: .circleci/run_on_circle.js [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+  console.log('body:', body);
```
Comment (manjeshbhargav) [inline .circleci/run_on_circle.js]:
```suggestion
```

File: .circleci/run_on_circle.js [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+  return new Promise((resolve, reject) => {
```
Comment (manjeshbhargav) [inline .circleci/run_on_circle.js]: @makarandp0 I've seen this pattern a few times now, so maybe we can refactor the code for making Circle CI API requests into its own module?

File: .circleci/run_on_circle.js [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    name: 'token',
+    message: 'Circle CI Token ?',
+    validate: (val) => { return typeof val === 'string' && val.length > 5; }
+  };
+
+  if (process.env.CIRCLE_CI_TOKEN) {
+    tokenPrompt.default = process.env.CIRCLE_CI_TOKEN;
+  }
+
+  inquirer.prompt([
```
Comment (manjeshbhargav) [inline .circleci/run_on_circle.js]: @makarandp0 I think we should be able to override all these options with a command line flag or an env variable. Then, if all options are provided, we can just show the confirm prompt at the end.

File: .circleci/run_on_circle.js [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    bverPrompt,
+    topologyPrompt,
+  ]).then(answers => {
+    const {options, body} = generateBuildRequest(answers);
+    console.log('Will make a CI request with:', options, body);
+    inquirer.prompt([confirmPrompt]).then(({ confirm }) => {
+      if (confirm) {
+        triggerBuild({options, body}).then((result) => {
+          console.log(result);
+        }).catch((e) => {
```
Comment (manjeshbhargav) [inline .circleci/run_on_circle.js]:
```suggestion
        }).catch(e => {
```

File: .circleci/run_on_circle.js [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+  };
+
+  const tokenPrompt = {
+    type: 'input',
+    name: 'token',
+    message: 'Circle CI Token ?',
+    validate: (val) => { return typeof val === 'string' && val.length > 5; }
+  };
+
+  if (process.env.CIRCLE_CI_TOKEN) {
```
Comment (manjeshbhargav) [inline .circleci/run_on_circle.js]:
```suggestion
  if (process.env.CIRCLECI_TOKEN) {
```

File: package.json [hunk 8]
```diff
@@ -7,7 +7,8 @@
   "author": "Mark Andrus Roberts <mroberts@twilio.com>",
   "contributors": [
     "Ryan Rowland <rrowland@twilio.com>",
-    "Manjesh Malavalli <mmalavalli@twilio.com>"
+    "Manjesh Malavalli <mmalavalli@twilio.com>",
+    "Makarand Patwardhan <mpatwardhan@twilio.com"
```
Comment (manjeshbhargav) [inline package.json]:
```suggestion
    "Makarand Patwardhan <mpatwardhan@twilio.com>"
```

File: .circleci/run_on_circle.js (hunk 5 above)
Comment (makarandp0) [reply .circleci/run_on_circle.js]: @manjeshbhargav - such command line processing would be useful when we invoke something from release tool. But for this tool it's useful to prompt everything - as this is essentially a replacement for the lack  of UI support on circle. I wanted to keep this code very simple.

File: .circleci/run_on_circle.js (hunk 4 above)
Comment (makarandp0) [reply .circleci/run_on_circle.js]: @manjeshbhargav - This `triggerBuild` method does not have anything circleCI specific -  it's just sending a request over.  CircleCI request specific logic is in `generateBuildRequest` - and that is tied closely to how we have structured our parameters in `config.yml`. For now this is the only place that does anything with circleCI request, I will look into making it reusable once we need it for release tool as well.

File: .circleci/run_on_circle.js (hunk 5 above)
Comment (makarandp0) [reply .circleci/run_on_circle.js]: but I agree, once we decide to use this tool from jenkins or other jobs, we would need command line options, and will add those as needed.

File: .circleci/runBuild.js [hunk 9]
```diff
@@ -0,0 +1,162 @@
+#!/usr/bin/env node
```
Comment (manjeshbhargav) [inline .circleci/runBuild.js]: @makarandp0 Can it be `runbuild.js`? We don't use camel-case for filenames either.

**Pull request merged**